src/sentry/api/base.py

```diff
@@ -404,6 +407,148 @@ def paginate(
         return response
 
 
+@dataclass
+class _VersionedMethodBody:
+    """Wraps an object method that handles versioned requests to an endpoint."""
+
+    http_method: str
+    version: int
+    wrapped: Callable
+
+    def __call__(self, *args: Any, **kwargs: Any) -> Any:
+        return self.wrapped(*args, **kwargs)
+
+
+def method_version(http_method: str, version: int = 0):
+    """Decorate an endpoint method to handle versioned requests.
+
+    This decorator should be used only on endpoint classes that extend
+    `VersionedEndpoint`. The default version (0) should be applied to methods
+    that existed before the platform supported versioning.
+
+    Example usage:
+
+        class AnEndpoint(VersionedEndpoint):
+            @method_version("get")
+            def get_v0(self, request):
+                return get_stuff(request)
+
+            @method_version("get", 1)
+            def get_v1(self, request):
+                return get_other_stuff(request)
+    """
+
+    if http_method not in APIView.http_method_names:
+        raise ValueError(f"http_method must be one of: {APIView.http_method_names}")
+
+    def decorator(wrapped: Callable) -> _VersionedMethodBody:
+        if wrapped.__name__ in APIView.http_method_names:
+            raise Exception(
+                "Methods decorated with @method_version must not "
+                f"override HTTP handler methods such as `{wrapped.__name__}`. "
+                "Give it a descriptive name or change it to "
+                f"`{wrapped.__name__}_v{version}`."
```

Comment:
What happens if the endpoint has both `get` and versioned methods? Will the versioned methods not be used?

Reply:
As of your comment, yes, it would have silently ignored the versioned methods.

As of a98edddb624d786a4e79007bcaf518ed30095019, that would raise a validation error.